core/contracts/Registry.sol

```diff
@@ -57,58 +55,84 @@ contract Registry is RegistryInterface, MultiRole {
     }
 
-    function getRegisteredDerivatives(address party) external view returns (address[] memory derivatives) {
-        // This is not ideal - we must statically allocate memory arrays. To be safe, we make a temporary array as long
-        // as registeredDerivatives. We populate it with any derivatives that involve the provided party. Then, we copy
-        // the array over to the return array, which is allocated using the correct size. Note: this is done by double
-        // copying each value rather than storing some referential info (like indices) in memory to reduce the number
-        // of storage reads. This is because storage reads are far more expensive than extra memory space (~100:1).
-        address[] memory tmpDerivativeArray = new address[](registeredDerivatives.length);
-        uint outputIndex = 0;
-        for (uint i = 0; i < registeredDerivatives.length; i = i.add(1)) {
-            address derivative = registeredDerivatives[i];
-            if (derivativesToParties[derivative].parties[party]) {
-                // Copy selected derivative to the temporary array.
-                tmpDerivativeArray[outputIndex] = derivative;
-                outputIndex = outputIndex.add(1);
+    function getRegisteredDerivatives(address party) external view returns (address[] memory) {
+        return partyMembersToDerivatives[party];
+    }
+
+    function addPartyToDerivative(address party) external {
+        address derivativeAddress = msg.sender;
+
+        require(
+            addressToDerivatives[derivativeAddress].valid == DerivativeValidity.Valid,
+            "Can add to valid derivative"
+        );
+        require(addressToDerivatives[derivativeAddress].parties[party] == false, "Can only register a party once");
+
+        addressToDerivatives[derivativeAddress].parties[party] = true;
+        partyMembersToDerivatives[party].push(derivativeAddress);
+
+        emit PartyMemberAdded(derivativeAddress, party);
+    }
+
+    function removedPartyFromDerivative(address party) external {
+        address derivativeAddress = msg.sender;
+
+        require(
+            addressToDerivatives[derivativeAddress].valid == DerivativeValidity.Valid,
+            "Remove only from valid derivative"
+        );
+        require(addressToDerivatives[derivativeAddress].parties[party] == true, "Remove existing party only");
+
+        addressToDerivatives[derivativeAddress].parties[party] = false;
+
+        // Need to delete the derivative from the partymembers array. This is vulnerable to a party member not being able
+        // to remove a derivative from their array, if they have too many in their array and exceed gas limit.
+        // However, this dos attack will not affect any party member other than the one who created too many.
+
+        // Deleting works by looping through all derivatives the party member has in their array until the location is
```

Comment:
This makes me pretty nervous -- is there a way we can do this in constant time?

My primary issue is something like the following scenario:

Let's assume some contract automatically adds and removes parties as users enter and exit.
Let's also assume that a keeper may run certain methods that cause *others* to be added and removed as a biproduct.
I can disrupt the keeper by making myself a party in many, many contracts, causing the keeper's transaction to run out of gas whenever it affects my position in the contract. 

I'm not sure any contract would actually work this way, but, in general, it seems bad to have an unbounded `for` loop like this running as a part of normal on-chain operations.

Reply:
This is a good point. I will refactor this to cover this constant time lookup.